fcomm/src/lib.rs

```diff
@@ -0,0 +1,1319 @@
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct VerificationResult {
+    pub verified: bool,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Proof<'a, F: LurkField> {
+    pub claim: Claim<F>,
+    pub proof: nova::Proof<'a, Coproc<S1>>,
+    pub num_steps: usize,
+    pub reduction_count: ReductionCount,
+}
+
+#[cfg_attr(not(target_arch = "wasm32"), derive(Arbitrary))]
+#[cfg_attr(not(target_arch = "wasm32"), proptest(no_bound))]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub enum Claim<F: LurkField> {
+    Evaluation(Evaluation),
+    // TODO: Add Expression type
+    PtrEvaluation(PtrEvaluation<F>),
+    Opening(Opening<F>),
+}
+
+impl<F: LurkField + Serialize + for<'de> Deserialize<'de>> Claim<F> {
+    // Returns the ZPtr corresponding to the claim
+    pub fn proof_key(&self) -> Result<ZExprPtr<F>, Error> {
+        match self {
+            Claim::Evaluation(eval) => {
+                // Only keying on input and output for now
+                let expr_in = ZExprPtr::<F>::from_lurk_str(&eval.expr)?;
+                let expr_out = ZExprPtr::<F>::from_lurk_str(&eval.expr_out)?;
+                let expr = ZExpr::Cons(expr_in, expr_out);
+                Ok(expr.z_ptr(&PoseidonCache::default()))
+            }
+            Claim::PtrEvaluation(ptr_eval) => {
+                let expr_in: ZExprPtr<F> = match &ptr_eval.expr {
+                    LurkPtr::Source(source) => ZExprPtr::<F>::from_lurk_str(source)?,
+                    LurkPtr::ZStorePtr(zsp) => zsp.z_ptr,
+                };
+                let expr_out = match &ptr_eval.expr_out {
+                    LurkPtr::Source(source) => ZExprPtr::<F>::from_lurk_str(source)?,
+                    LurkPtr::ZStorePtr(zsp) => zsp.z_ptr,
+                };
+                let expr = ZExpr::Cons(expr_in, expr_out);
+                Ok(expr.z_ptr(&PoseidonCache::default()))
+            }
+            // TODO: Is this an appropriate key for commitments?
+            Claim::Opening(open) => {
+                let expr_in = ZExprPtr::<F>::from_lurk_str(&open.input)?;
+                let expr_out = ZExprPtr::<F>::from_lurk_str(&open.output)?;
+                let expr = ZExpr::Cons(expr_in, expr_out);
+                Ok(expr.z_ptr(&PoseidonCache::default()))
+            }
+        }
+    }
+}
+
+// This is just a rough idea, mostly here so we can plumb it elsewhere. The idea is that a verifier can sign an
+// attestation that a given claim's proof was verified. It motivates the use of an online verifier for demo purposes.
+// Although real proofs should be fast to verify, they will still be large relative to a small (auditable) bundle like
+// this. Even if not entirely realistic, something with this general *shape* is likely to play a role in a recursive
+// system where the ability to aggregate proof verification more soundly is possible.
+//#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+//pub struct Cert {
```

Comment:
This can be uncommented, it's not used at all but may be useful in a future `fcomm` application. cc @porcuquine

Reply:
I would uncomment this, but I think the way we're serializing things has changed since this was implemented (as it refers explicitly to IPLD cids in the implementation). It may be worth keeping this commented and add an issue to uncomment and bring it in line with the current serde implementations.

Reply:
Technically, this is actually [used](https://github.com/lurk-lab/zaps/blob/da582b650d06be4b4acf064f50e4b09ab1a1d570/src/fcomm.rs#L70-L76) by the [zaps repo](https://github.com/lurk-lab/zaps), which I think is also itself used by our [website](https://www.lurk-lang.org/). (cc: @samuelburnham). I mention this mainly so we remember these downstream dependencies exists. As a matter of principle, we should probably work to retain functionality. Whether we prioritize updating how everything works or not, we should minimize the complexity of doing so if we want to. We should probably, minimally, update web examples to be fully compatible at some point down the road.

Reply:
@mpenciak this PR is supposed to move code from where it is on master (in `public_parameters`) to where it should be (in `fcomm`). The snippet in question exists in adapted, uncommented form in master, and you're removing it here:
https://github.com/lurk-lab/lurk-rs/pull/447/files#diff-372f8376bdf97a030ac402c26c94fd3ba3ee64fb207d20c2ff7c446f0a4bd916L324-L336
So we should move that adapted, uncommented form here.